Add an example for TensorFlow Decision Forests.

ksalama opened the pull request:
Hello @fchollet -

Kindly review this PR that provides an example of using [TensorFlow Decision Forests](https://www.tensorflow.org/decision_forests) for structured data classification. The example covers three scenarios:

1. Build a Gradient Boosted Trees model by specifying the input feature usage.
2. Implement a custom **Binary Target encoder** as a Keras Preprocessing layer to encode the categorical features with respect to their target value co-occurrences.
3. Encode the categorical features as embeddings, train these embeddings in a simple linear model, and then use the trained embeddings as inputs to build decision forests model. 

As for the second scenario, this [blog post](https://medium.com/analytics-vidhya/target-encoding-vs-one-hot-encoding-with-simple-examples-276a7e7b3e64) provided a simple explanation to Target Encoding. I think it is valuable to show an example of how to create a custom preprocessing layer that required and `adapt` phase, similar to the Keras [Normalization](https://keras.io/api/layers/preprocessing_layers/core_preprocessing_layers/normalization/) preprocessing layer. 

In fact, I tried to follow the [source code for Normalization](https://github.com/keras-team/keras/blob/v2.7.0/keras/layers/preprocessing/normalization.py#L30-L279) by just overriding the `update_state`, `reset_state` and `finalize_state` methods, but I couldn't make it working with `iterator` that is passed to these methods. Thus, I had to override the `adapt` method itself.

It would be great If you can advise on the best way to implement this. Note that in Decision Forests, we expect all the data to be available in memory, thus we don't need to support reading batches and updating the state of the preprocessor (although it should be quite easy to do so in Target Encoding).

One last thing. I know that the example is a little bit long. But once we review/improve/approve the target encoding thing, we can decide on how we can reduce its size, or split it into multiple examples (maybe a recipe and and example, etc.).

Thank you in advance for your help

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+"""
+## Define dataset metadata
+
+Here, we define the metadata of the dataset that will be useful for encoding
```
What about saying that, with Keras's TF-DF can consume string feature values natively. And those will, by default, be considered categorical values. In other terms, there is no need for the user to specify the dictionary.

The categorical feature dictionaries are used in the TargetMean and embedding section, so I would compute them here.

ksalama replied:
In fact, I mentioned how TF-DF can infer the semantic and consume (string) feature values natively automatically later in the example. I prefer not to provide a lot if details here. I am happy to remove this text that says "why" we need the metadata.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+"""
+
+# Target feature name.
+TARGET_FEATURE_NAME = "income_level"
```
s/TARGET_FEATURE_NAME/LABEL_KEY or s/TARGET_FEATURE_NAME/TARGET_COLUMN_NAME (or something like that).

The label is a column, but calling it a "feature" feels weird.

ksalama replied:
Agreed

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+# Categorical features and their vocabulary lists.
+CATEGORICAL_FEATURES_WITH_VOCABULARY = {
+    feature_name: sorted(
+        [str(value) for value in list(train_data[feature_name].unique())]
```
I would cast all the bytes feature into str when loading the dataset to avoid having to handle both byte and str objects here and after.

What about creating a "process_dataframe" function that does:
- The conversion of byte to str of string values.
- The conversion from str to integer of the label values.
- ... any other dataset preprocessing applied before the tensorflow code.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+[documentation](https://www.tensorflow.org/decision_forests/api_docs/python/tfdf/keras/GradientBoostedTreesModel)
+"""
+
+GROWING_STRATEGY = "BEST_FIRST_GLOBAL"
```
I would keep the default hyper-parameters of the algorithm (i.e. don't specify any hyper-parameters at first).

If hyper-parameters are tuned, I would make sure to explain why and to have the colab print evaluation results.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2022/01/31
+Last modified: 2022/01/31
+Description: Using TensorFlow Decision Forests for structured data classification.
```
Nit: This colab is about two advanced ideas: Target mean and composition of Linear embedding and Decision Forests. Should the title reflect this?

Also, what about linking the TF-DF introductory colab (https://www.tensorflow.org/decision_forests/tutorials/beginner_colab), and mention that it is a recommended pre-requirement for this colab?

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+"""
+
+
+def fix_datatypes(features):
```
Merge with the `process_dataframe` function discussed above.

fchollet replied:
The calls to `map(lambda features, target, weight: (fix_datatypes(features), target, weight))` are heavy to read, so I recommend turning this into a function with signature `prepare_sample(features, target, weight)`

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+    return features
+
+
+def run_experiment(model, train_data, test_data, num_epochs=1, batch_size=None):
```
Add comment on what this function is doing.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+# Target feature name.
+TARGET_FEATURE_NAME = "income_level"
+# Weight column name.
+WEIGHT_COLUMN_NAME = "instance_weight"
```
In this dataset, the "instance_weight" is not a weight to use to train a classifier. See comment in dataset readme:

```
| The instance weight indicates the number of people in the population
| that each record represents due to stratified sampling.
| To do real analysis and derive conclusions, this field must be used.
| This attribute should *not* be used in the classifiers, so it is
| set to "ignore" in this file.
```

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+        test_data, label=TARGET_FEATURE_NAME, weight=WEIGHT_COLUMN_NAME
+    ).map(lambda features, target, weight: (fix_datatypes(features), target, weight))
+
+    history = model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
```
No need for num_epoch and batch_size.

You can add comment that the algorithm should be trained for exactly one epoch and that the batch size does not matter.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+
+def create_model_inputs():
+    inputs = {}
+    for feature_name in FEATURE_NAMES:
```
inputs = [ layers.Input(
                name=name, shape=(), dtype=tf.float32 if name in NUMERIC_FEATURE_NAMES else tf.string
            ) for name in FEATURE_NAMES]

Same for `specify_feature_usages`.

Question: What about the categorical features stored as integers e.g. `detailed_industry_recode`?

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+        exclude_non_specified_features=True,
+        growing_strategy=GROWING_STRATEGY,
+        num_trees=NUM_TREES,
+        max_depth=MAX_DEPTH,
```
Remove:
```
        growing_strategy=GROWING_STRATEGY,
        num_trees=NUM_TREES,
        max_depth=MAX_DEPTH,
        min_examples=MIN_EXAMPLES,
        subsample=SUBSAMPLE,
        validation_ratio=VALIDATION_RATIO,
        task=tfdf.keras.Task.CLASSIFICATION,
        loss="DEFAULT",
```

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+            True
+        )
+
+    def adapt(self, data):
```
With pandas, Target mean can be done in two lines:

```python
ds = pd.DataFrame(
    {"cat_feature":["a","b","c","a","b","c"],
     "binary_label":[0,0,1,0,1,1]})

target_map = ds.groupby("cat_feature")["binary_label"].mean()
ds["taget_mean_feature"] = ds["cat_feature"].map(target_map)
```

Maybe something similar could be done here. Note that group_by is [available](https://www.tensorflow.org/api_docs/python/tf/data/experimental/group_by_reducer) in tf.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+        feature_values = data[:, 0]
+        target_values = data[:, 1]
+        # Validate that the target has only 0 and 1 values.
+        unique_target_values = tf.sort(tf.unique(target_values).y)
```
I think tf.sort and unique are done in memory.
It would be interesting to understand and document what append in this processing layer when used on a large dataset. For example, will the entire dataset be loaded in memory?

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+        )
+        # Compute positive probability for the input feature values.
+        positive_probability = positive_fequency / (
+            positive_fequency + negative_fequency
```
This division will be unstable for rare categorical values. With target mean, it works well to add a correction term controlled by an hyperparameter.

```python
correction =  1
positive_probability = positive_frequency / (positive_frequency + negative_frequency + correction)
```

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+            dtype=tf.dtypes.float32,
+        )
+        # Lookup negative frequencies for the input feature values.
+        negative_fequency = tf.cast(
```
s/fequency/frequency

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+
+    def compute_output_signature(self, input_spec):
+        output_shape = self.compute_output_shape(input_spec.shape.as_list())
+        return tf.TensorSpec(shape=output_shape, dtype=tf.int64)
```
There is some mixup with precision between the call and this method.

I would simply return a float32 for both the frequencies and ratios.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+        # Lookup negative frequencies for the input feature values.
+        negative_fequency = tf.cast(
+            tf.gather_nd(self.negative_frequency_lookup, inputs),
+            dtype=tf.dtypes.float32,
```
Note that this is a number computed possible over many examples. So float32 precision might not be suited.

ksalama replied:
Oh I just need to cast to float32 as I am concatenating it with `positive_probability` later, which is float32: 
```    
return layers.concatenate(
   [positive_fequency, negative_fequency, positive_probability], axis=1)
```

achoum replied:
It is probably a nit given that this is an example. But while, computation on a single example (or a batch of examples) is generally fine in float32. Computation on many examples (possibly a large part of the datasets) is better done in double.

In this example, I would do the division in double (float64) and then cast the result to float32.

ksalama replied:
Oh I see. Thank you for the comment!

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+    target_values = train_data[[TARGET_FEATURE_NAME]].to_numpy()
+    encoded_features = []
+    for feature_name in inputs:
+        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
```
Target mean should also be used on categorical feature stored as integers.

ksalama replied:
`CATEGORICAL_FEATURES_WITH_VOCABULARY` include both categorical features with string values and integer values.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+
+embedding_encoder = create_embedding_encoder()
+history = run_experiment(
+    create_linear_model(embedding_encoder),
```
This is not really an embedding. Instead, it will be the output of the linear model trained with categorical buckets. So it is fundamentally very similar to target mean.

What about having a neural networks with 1 or 2 hidden layers instead? Note that in this case, the input of the DF should not be the output of the neural networks. Instead, it should be the connected to one of the hidden layers.

See https://www.tensorflow.org/decision_forests/tutorials/model_composition_colab

ksalama replied:
I am not entirely sure why you think these are not embeddings (and what you mean by categorical buckts)? 


The embedding encoder create a dense vector for each categorical value (which for example could be of size 10 dims if the vocabulary size is 100), and these values of these dense vectors are learnt via backprop while training the linear model. 

Not that the input to the DF is not the output of the linear model, rather the embedding encoder that converts the categorical features to embeddings.

 On the other hand the `Binary Target Encoder` produced exactly three features for each categorical feature (regardless of the vocabulary size of the categorical feature). These new features are not trained, rather computed as statistics about co-occurrence of the feature and the target values.

I am really sorry if there is something that is not clear to me, but I am here to listen and learn :-)

achoum replied:
>> I am not entirely sure why you think these are not embeddings (and what you mean by categorical buckts)?

buckts -> buckets.

Long story short: Because there are no activation functions in between your embedding layer and the only hidden layer of your model.

**Details**

There are no activation functions in between your embedding layer and the only hidden layer of your model.

(I think that) in this current setup, the network is equivalent (but more expensive) to a linear model where all the categorical features are feed with one-hot. In other words, in the best case (i.e. the network is fully and correctly trained), this network will be some kind of a Naive Bayes classifier i.e. each possible categorical value is treated independently.

When the DF receives the intermediate value between the embedding layer and the hidden linear layer, and because there are not activation in between those layers, the DF essentially receives a mix of random linear projection (because here, the repartition of the signal between the embedding and the hidden layer is random [since there is not part of the loss that push it on one side or the other]. Random linear projections are sometime very good, but this is not what you want here: In this example, you want to feed the output of an embedding.

To solve this issue, you can add an activation to the embedding, or equivalently, add more hidden layers to the network.

>> Feeding of linear model.

My bad. I through the output of the linear model was feed into the DF directly.
Instead, you are feeding the output of the embedding directly.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+    create_linear_model(embedding_encoder),
+    train_data,
+    test_data,
+    num_epochs=3,
```
The number of epoch only make sense to the linear or embedding layer.

ksalama replied:
Yes. That is why I am only passing num_epochs when training the linear model, not when training a DF model (default is 1). Am I missing something?

I added the following text to clarify that in the "train and evaluate the model" step of the first experiment:

> Note that when training a Decision Forests model, only one epoch is needed to
> read the full dataset. Any extra steps will result in unnecessary slower training.
> Therefore, the default `num_epochs=1` is used in the `run_experiment` method.

achoum replied:
I see. So the code is equivalent to the following one, and batch_size=None will be ignored by the Keras API.

```python
num_epochs = 1
batch_size = None

model = tfdf.keras.RandomForest()
model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
```

So the code is correct. It just feel a bit weird since specifying the number of epoch does not make sense for the TF-DF model, and readers might miss that part i.e. they might try to run the num_epoch.

Nit: What about having a **fit_args argument to the `run_experiment` function?

ksalama replied:
Sounds good.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+for their training procedures. However, there are diminishing returns
+for increasing the size of the dataset, and Decision Forests algorithms
+arguably need fewer examples for convergence than large Neural Network models.
+"""
```
Could the colab report the test accuracy of the three tested methods?

Example: https://www.tensorflow.org/decision_forests/tutorials/model_composition_colab#whats_next

ksalama replied:
It does. In the `run_experiment` function, the test accuracy is printed:

```
    history = model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
    _, accuracy = model.evaluate(test_dataset, verbose=0)
    print(f"Test accuracy: {round(accuracy * 100, 2)}%")
```

Am I missing something?

achoum replied:
I don't see the output of the colab in this review.

Also, it might be easier for the reader to re-print together all the accuracies of the models in a single cell at the end.

In the colab I lnked, all the accuracies are re-printed at the end as follow:

```
Accuracy (model #1): 0.820300
Accuracy (model #2): 0.821800
Accuracy (model #3): 0.822300
----------------------------------------
Accuracy (model #3): 0.823600
                  +0.003300 over model #2
                  +0.001800 over model #1
```

Then, you could write a paragraph that describes the observations and conclusions to help the reader. For example, I suppose that the accuracy with the embedding or the target mean will be better than the accuracy of the initial model (otherwise, why going over the troubles?).

ksalama replied:
Done. I added the results in the concluding remarks section

achoum reviewed (commented):
Great colab :).

A few comments and nits.

ksalama commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+]
+
+train_data = pd.read_csv(
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.data.gz",
```
I used to this, but I am limited with the number of code lines in the example, so I started to reduce the code lines :-)

If we decide to split this into two examples (maybe one recipe covering Target Encoding) and other showing the first and the last experiments, I'll go ahead and add these lines

ksalama replied:
Done

ksalama commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+"""
+
+# Column names.
+CSV_HEADER = [
```
Done

ksalama commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,726 @@
+We convert the target column from string to integer.
+"""
+
+train_data["income_level"] = train_data["income_level"].apply(
```
Done

ksalama commented:
@achoum - Thank you so much for your comments. I fixed the code accordingly, and provided responses to comments.

Could you also provide some advice around the implementation of the BinaryTargetEncoder, and if you think it is better to split the example into two: 1) a recipe with the Target encoder and covert experiment 2, and 2) a separate example to cover experiment 1 and 3? Cheers

BTW, I am a big fan of TF-DF! Great work 👍

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,682 @@
+    ).map(lambda features, target, weight: (fix_datatypes(features), target, weight))
+    test_dataset = tfdf.keras.pd_dataframe_to_tf_dataset(
+        test_data, label=TARGET_COLUMN_NAME, weight=WEIGHT_COLUMN_NAME
+    ).map(lambda features, target, weight: (fix_datatypes(features), target, weight))
```
Calls to `map()` need a `num_parallel_calls`

ksalama replied:
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,682 @@
+    _, accuracy = model.evaluate(test_dataset, verbose=0)
+    print(f"Test accuracy: {round(accuracy * 100, 2)}%")
+
+    return history
```
If there's only one epoch I guess history is not useful

ksalama replied:
Makes sense :-)

Removed

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,682 @@
+### Implement Binary Target Encoder
+"""
+
+from keras.engine import base_preprocessing_layer
```
This is a private API for now, do not import it. Create you layer by subclassing `Layer`.

ksalama replied:
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,682 @@
+            [positive_fequency, negative_fequency, positive_probability], axis=1
+        )
+
+    def compute_output_shape(self, input_shape):
```
A lot of the logic in the layer should not be necessary here, including this method and the ones below.

ksalama replied:
Ok removed

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,682 @@
+from keras.engine import base_preprocessing_layer
+
+
+class BinaryTargetEncoding(base_preprocessing_layer.PreprocessingLayer):
```
What does this layer do? Without reading the code in detail, it sounds like it would be the same as `CategoricalEncoding`.

ksalama replied:
Given a categorical feature, the `BinaryTargetEncoding` will produce three new numerical features:

1. `positive_frequency`: How many times each feature value occurred with a positive target label.
2. `negative_frequency`: How many times each feature value occurred with a negative target label.
3. `positive_probability`: The probability that the target label is positive,
given the feature value, which is computed as
`positive_frequency / (positive_frequency + negative_frequency)`.

So `BinaryTargetEncoding` uses the categorical feature values and the target values to encode the categorical feature.

fchollet reviewed (commented):
Thanks for the PR! Demonstrating TF-DF workflows is super useful 👍

ksalama commented on examples/structured_data/classification_with_tfdf.py:
````diff
@@ -0,0 +1,682 @@
+as well as [TensorFlow Decision Forests](https://www.tensorflow.org/decision_forests),
+which you can install using the following command:
+
+``` python
````
Done

ksalama commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,682 @@
+
+print(f"Train data shape: {train_data.shape}")
+print(f"Test data shape: {test_data.shape}")
+train_data.head().T
```
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,669 @@
+            [positive_fequency, negative_fequency, positive_probability], axis=1
+        )
+
+    def _standardize_inputs(self, inputs):
```
Remove this method and instead inline a call to `tf.cast` in the 2 places where it was used (the rest of the logic in the function is redundant)

ksalama replied:
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,669 @@
+        inputs = self._standardize_inputs(inputs)
+
+        # Fix and validate inputs shape.
+        if inputs.shape.rank == 1:
```
You could remove this block and just require rank 2, this is a code example and it needs to be kept short & focused

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,669 @@
+            positive_fequency + negative_fequency
+        )
+        # Concatenate and return the looked-up statistics.
+        return layers.concatenate(
```
Just `tf.concat`

ksalama replied:
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -0,0 +1,669 @@
+        # data is expected to be an integer numpy array to a Tensor shape [num_exmples, 2].
+        # This contains feature values for a given feature in the dataset, and target values.
+
+        self.compile()
```
?

fchollet reviewed (commented):
Thanks for the update!

ksalama commented:
Hello @fchollet - I think I addressed most of your comments. Also, now the example is under 300 lines, so no need for splitting it. LMK if you have any further comments before generating the other files

fchollet reviewed (approved):
Awesome! Please add the generated files.

fchollet commented:
Thank you for the great contribution!
